Sparse CSR: Add `torch.randn_like`

IvanYashchuk opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #68084
* __->__ #68083
* #71103

This PR adds support for `torch.randn_like(sparse_csr_tensor)`.
It creates a new sparse csr tensor with same indices but different values that are normally distributed.

In addition `.normal_()` and `torch.empty_like` were implemented because `randn_like` is a composite of these two functions.

cc @nikitaved @pearu @cpuhrsch @IvanYashchuk

Differential Revision: [D33511280](https://our.internmc.facebook.com/intern/diff/D33511280)

IvanYashchuk committed:
Sparse CSR: Add `torch.randn_like`

This PR adds support for `torch.randn_like(sparse_csr_tensor)`.
It creates a new sparse csr tensor with same indices but different values that are normally distributed.

In addition `.normal_()` and `torch.empty_like` were implemented because `randn_like` is a composite of these two functions.

[ghstack-poisoned]

IvanYashchuk commented on test/test_sparse_csr.py:
```diff
@@ -1070,6 +1074,23 @@ def test_coo_csr_conversion(self, device, dtype):
+            expected = op(sample.input)
+            assert torch.is_tensor(expected)
+            output = op(sample.input.to_sparse_csr())
+            assert torch.is_tensor(output)
```
We can actually compare `expected` and `output.to_dense()` with `self.assertEqual`. It passes for `randn_like`, but of course fails for `empty_like`.

pearu replied:
> It passes for randn_like.

I think this happens by accident (if `sample.input` is a random dense tensor it likely does not contain zeros, that is, it is full tensor, then also `to_sparse_csr` results in a full tensor) and not in general (if `sample.input` would not be a full tensor then `assertEqual(output.to_dense(), expected)` will fail at unspecified element indices).

IvanYashchuk replied:
It was a carefully designed test of creating a sparse representation of a surely dense tensor, so not by accident 😄

pearu replied:
Not sure what do you mean by a carefully designed test considering that the domain of a number of operations that support sparse csr inputs, contains 0.

krshrimali commented on aten/src/ATen/native/sparse/SparseCsrTensorMath.cpp:
```diff
@@ -59,6 +59,14 @@ void convert_indices_from_coo_to_csr_cpu(const Tensor& result, const Tensor& inp
+template <typename F, typename ...Args>
+Tensor& unary_inplace_op_(F op_, Tensor& self, Args&&... args) {
+  TORCH_INTERNAL_ASSERT(self.is_sparse_csr());
+  auto values = self.values();
+  (self.values().*op_)(std::forward<Args>(args)...);
```
Do we need `auto values = self.values()` here? I feel it's redundant?

krshrimali commented on aten/src/ATen/native/TensorFactories.cpp:
```diff
@@ -272,6 +272,18 @@ Tensor empty_like(
+        self.col_indices().clone(),
+        at::empty(self.values().sizes(), options.layout(kStrided)),
+        self.sizes(),
+        self.scalar_type(),
```
Post offline discussion with @IvanYashchuk - This should be `dtype` if provided, else `self.scalar_type()`.

IvanYashchuk replied:
This should be `options.dtype()`, the device should also be obtained from `options`.

IvanYashchuk committed:
Update on "Sparse CSR: Add `torch.randn_like`"


This PR adds support for `torch.randn_like(sparse_csr_tensor)`.
It creates a new sparse csr tensor with same indices but different values that are normally distributed.

In addition `.normal_()` and `torch.empty_like` were implemented because `randn_like` is a composite of these two functions.

cc nikitaved pearu cpuhrsch IvanYashchuk

[ghstack-poisoned]

IvanYashchuk committed:
Update on "Sparse CSR: Add `torch.randn_like`"


This PR adds support for `torch.randn_like(sparse_csr_tensor)`.
It creates a new sparse csr tensor with same indices but different values that are normally distributed.

In addition `.normal_()` and `torch.empty_like` were implemented because `randn_like` is a composite of these two functions.

cc nikitaved pearu cpuhrsch IvanYashchuk

[ghstack-poisoned]

IvanYashchuk committed:
Update on "Sparse CSR: Add `torch.randn_like`"


This PR adds support for `torch.randn_like(sparse_csr_tensor)`.
It creates a new sparse csr tensor with same indices but different values that are normally distributed.

In addition `.normal_()` and `torch.empty_like` were implemented because `randn_like` is a composite of these two functions.

cc nikitaved pearu cpuhrsch IvanYashchuk

[ghstack-poisoned]

IvanYashchuk committed:
Update on "Sparse CSR: Add `torch.randn_like`"


This PR adds support for `torch.randn_like(sparse_csr_tensor)`.
It creates a new sparse csr tensor with same indices but different values that are normally distributed.

In addition `.normal_()` and `torch.empty_like` were implemented because `randn_like` is a composite of these two functions.

cc nikitaved pearu cpuhrsch IvanYashchuk

[ghstack-poisoned]

IvanYashchuk committed:
Update on "Sparse CSR: Add `torch.randn_like`"


This PR adds support for `torch.randn_like(sparse_csr_tensor)`.
It creates a new sparse csr tensor with same indices but different values that are normally distributed.

In addition `.normal_()` and `torch.empty_like` were implemented because `randn_like` is a composite of these two functions.

cc nikitaved pearu cpuhrsch IvanYashchuk

[ghstack-poisoned]

IvanYashchuk committed:
Update on "Sparse CSR: Add `torch.randn_like`"


This PR adds support for `torch.randn_like(sparse_csr_tensor)`.
It creates a new sparse csr tensor with same indices but different values that are normally distributed.

In addition `.normal_()` and `torch.empty_like` were implemented because `randn_like` is a composite of these two functions.

cc nikitaved pearu cpuhrsch IvanYashchuk

[ghstack-poisoned]

IvanYashchuk committed:
Update on "Sparse CSR: Add `torch.randn_like`"


This PR adds support for `torch.randn_like(sparse_csr_tensor)`.
It creates a new sparse csr tensor with same indices but different values that are normally distributed.

In addition `.normal_()` and `torch.empty_like` were implemented because `randn_like` is a composite of these two functions.

cc nikitaved pearu cpuhrsch IvanYashchuk

[ghstack-poisoned]

cpuhrsch commented:
On a high level note: In light of our recent decisions around sparsity as transparent layouts and constraining unspecified values to mean zero, how do we best fit operators like randn_like or normal into it?

pearu commented on test/test_sparse_csr.py:
```diff
@@ -1070,6 +1074,23 @@ def test_coo_csr_conversion(self, device, dtype):
+
+        for sample in samples:
+            assert isinstance(sample.input, torch.Tensor)
+            if sample.input.ndim != 2:
+                continue
```
With https://github.com/pytorch/pytorch/blob/6c4437118b88c5756b0a192263fb8ed40cfe2800/torch/testing/_internal/common_methods_invocations.py#L1184-L1186, this if-block is redudant.

IvanYashchuk committed:
Update on "Sparse CSR: Add `torch.randn_like`"


This PR adds support for `torch.randn_like(sparse_csr_tensor)`.
It creates a new sparse csr tensor with same indices but different values that are normally distributed.

In addition `.normal_()` and `torch.empty_like` were implemented because `randn_like` is a composite of these two functions.

cc nikitaved pearu cpuhrsch IvanYashchuk

[ghstack-poisoned]

IvanYashchuk committed:
Update on "Sparse CSR: Add `torch.randn_like`"


This PR adds support for `torch.randn_like(sparse_csr_tensor)`.
It creates a new sparse csr tensor with same indices but different values that are normally distributed.

In addition `.normal_()` and `torch.empty_like` were implemented because `randn_like` is a composite of these two functions.

cc nikitaved pearu cpuhrsch IvanYashchuk

[ghstack-poisoned]

IvanYashchuk committed:
Update on "Sparse CSR: Add `torch.randn_like`"


This PR adds support for `torch.randn_like(sparse_csr_tensor)`.
It creates a new sparse csr tensor with same indices but different values that are normally distributed.

In addition `.normal_()` and `torch.empty_like` were implemented because `randn_like` is a composite of these two functions.

cc nikitaved pearu cpuhrsch IvanYashchuk

Differential Revision: [D33511280](https://our.internmc.facebook.com/intern/diff/D33511280)

[ghstack-poisoned]

IvanYashchuk committed:
Update on "Sparse CSR: Add `torch.randn_like`"


This PR adds support for `torch.randn_like(sparse_csr_tensor)`.
It creates a new sparse csr tensor with same indices but different values that are normally distributed.

In addition `.normal_()` and `torch.empty_like` were implemented because `randn_like` is a composite of these two functions.

cc nikitaved pearu cpuhrsch IvanYashchuk

Differential Revision: [D33511280](https://our.internmc.facebook.com/intern/diff/D33511280)

[ghstack-poisoned]

cpuhrsch commented:
@cpuhrsch has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33511280).
